testing/async/mixin.h

```diff
@@ -0,0 +1,645 @@
+//    // TEST_FUTURE_VALUE_RETRIEVAL), then check that the reference output
+//    // matches the testing output.
+//    template <typename EventType>
+//    static void compare_outputs(EventType &e,
+//                                output_type const &ref,
+//                                output_type const &test)
+//    {
+//      TEST_EVENT_WAIT(e);
+//      ASSERT_EQUAL_QUIET(ref, test);
+//    }
+// };
+// ```
+//
+// Similar invokers with slight tweaks are used in other
+// async/exclusive_scan/*.cu tests.
+
+namespace testing
+{
+namespace async
+{
+namespace mixin
+{
+
+//------------------------------------------------------------------------------
+namespace input
+{
+
+// TODO it'd be nice to specify a lambda expression that'd replace the call to
+// thrust::sequence when desired.
+template <typename value_type>
+struct device_vector
+{
+  using input_type = thrust::device_vector<value_type>;
+
+  static input_type generate_input(std::size_t num_values)
+  {
+    input_type input(num_values);
+    thrust::sequence(input.begin(),
+                     input.end(),
+                     static_cast<value_type>(1.5),
+                     static_cast<value_type>(1));
+    return input;
+  }
+};
+
+template <typename value_type>
+struct counting_iterator_0
+{
+  struct input_type
+  {
+    using iterator = thrust::counting_iterator<value_type>;
+
+    std::size_t num_values;
+
+    iterator begin() { return iterator{static_cast<value_type>(0)}; }
+    iterator begin() const { return iterator{static_cast<value_type>(0)}; }
+    iterator cbegin() const { return iterator{static_cast<value_type>(0)}; }
+
+    iterator end() { return iterator{static_cast<value_type>(num_values)}; }
+    iterator end() const { return iterator{static_cast<value_type>(num_values)}; }
+    iterator cend() const { return iterator{static_cast<value_type>(num_values)}; }
+
+    std::size_t size() const { return num_values; }
+  };
+
+  static input_type generate_input(std::size_t num_values)
+  {
+    return {num_values};
+  }
+};
+
+template <typename value_type>
+struct counting_iterator_1
+{
+  struct input_type
+  {
+    using iterator = thrust::counting_iterator<value_type>;
+
+    std::size_t num_values;
+
+    iterator begin() { return iterator{static_cast<value_type>(1)}; }
+    iterator begin() const { return iterator{static_cast<value_type>(1)}; }
+    iterator cbegin() const { return iterator{static_cast<value_type>(1)}; }
+
+    iterator end() { return iterator{static_cast<value_type>(1 + num_values)}; }
+    iterator end() const { return iterator{static_cast<value_type>(1 + num_values)}; }
+    iterator cend() const { return iterator{static_cast<value_type>(1 + num_values)}; }
+
+    std::size_t size() const { return num_values; }
+  };
+
+  static input_type generate_input(std::size_t num_values)
+  {
+    return {num_values};
+  }
+};
+
+template <typename value_type>
+struct constant_iterator_1
+{
+  struct input_type
```

Comment:
We should probably explain somewhere that the `input_type`s need random access (thus begin, end, and a size), while the `output_type`s (like discard) only need to satisfy output iterator / range.

Reply:
Added this info to the appropriate sections in the class documentation.